src/backend/optimizer/util/plancat.c

```diff
@@ -80,6 +82,28 @@ static void set_baserel_partition_key_exprs(Relation relation,
 static void set_baserel_partition_constraint(Relation relation,
 											 RelOptInfo *rel);
 
+static bool
+is_index_valid(Relation index)
+{
+	if (!index->rd_index->indisvalid)
+		return false;
+
+	if (index->rd_rel->relpersistence == RELPERSISTENCE_UNLOGGED)
+	{
+		Buffer metapage = ReadBuffer(index, 0);
+		bool isNew = PageIsNew(BufferGetPage(metapage));
+		ReleaseBuffer(metapage);
+		if (isNew)
+		{
+			Relation heap;
+			DropRelFileNodesAllBuffers(&index->rd_smgr, 1);
+			heap = RelationIdGetRelation(index->rd_index->indrelid);
+			index->rd_indam->ambuild(heap, index, BuildIndexInfo(index));
```

Comment:
> I'd like it if we didn't have to do an index build in the planner, even if the table is extremely unlikely to contain data at this point.

I also do not like it. But I do not know better solution. If error is reported during planning, then it should be fixed before it.
Certainly we can implement on-startup hook which will traverse pg_class, find all unlogged tables and reinitialize it.
But frankly speaking I do not think that it is better solution.

Reply:
> Also, could you check that this doesn't break on partitioned indexes?

Yes, I have checked it.  It works. And no surprise here because all metadata of unlogged tables is logged, so survive compute restart.